File: LeetCode/416.Partition_Equal_Subset_Sum/Mario/416.Partition_Equal_Subset_Sum.java
class Solution {
    public boolean canPartition(int[] nums) {
        int n = nums.length;
        int sum = 0;
        for (int num : nums)
            sum += num;
        if (sum % 2 == 1)
            return false;

        int t = sum / 2;
        int[] dp = new int[t + 1];

        for (int i = 0; i < n; i++) {
            for (int j = t; j >= nums[i]; j--) {
                dp[j] = Math.max(dp[j - nums[i]] + nums[i], dp[j]);
            }
            // System.out.println(Arrays.toString(dp));
        }

        return dp[t] == t;
    }
}

File: LeetCode/416.Partition_Equal_Subset_Sum/416.Partition_Equal_Subset_Sum.py
#!/usr/bin/env python3


class Solution:
    def canPartition(self, nums: List[int]) -> bool:
        # 0-1 knapsack
        # handle edge case
        total = sum(nums)
        if total % 2 == 1 or len(nums) <= 1:
            return False

        half = total // 2  # 11

        # 1 represent yes, 0 represent no
        # dp[i] can be sum i or not
        dp = [False for _ in range(half + 1)]
        dp[0] = True

        for num in nums:
            for j in range(half, num - 1, -1):
                dp[j] = dp[j] or dp[j - num]

        return dp[half]
